src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+            }
+        }
+
+        enum State
+        {
+            Stopped = -1,
+            Set = 0,
+            Clear = 1,
+        }
+
+        struct Queue<TOperation>
+            where TOperation : AsyncOperation
+        {
+            private AsyncOperation _tail;
+
+            public State State { get; set; }
+            public bool IsStopped { get { return State == State.Stopped; } }
+            public bool IsEmpty { get { return _tail == null; } }
+
+            public TOperation Head
+            {
+                get
+                {
+                    Debug.Assert(!IsStopped);
+                    return (TOperation)_tail.Next;
+                }
+            }
+
+            public TOperation Tail
+            {
+                get
+                {
+                    Debug.Assert(!IsStopped);
+                    return (TOperation)_tail;
+                }
+            }
+
+            public void Enqueue(TOperation operation)
+            {
+                Debug.Assert(!IsStopped);
+
+                if (!IsEmpty)
+                {
+                    operation.Next = _tail.Next;
+                    _tail.Next = operation;
+                }
+
+                _tail = operation;
+            }
+
+            public void Dequeue()
+            {
+                Debug.Assert(!IsStopped);
+                Debug.Assert(!IsEmpty);
+
+                AsyncOperation head = _tail.Next;
+                if (head == _tail)
+                {
+                    _tail = null;
+                }
+                else
+                {
+                    _tail.Next = head.Next;
+                }
+            }
+
+            public Queue<TOperation> Stop()
+            {
+                Queue<TOperation> result = this;
+                _tail = null;
+                State = State.Stopped;
+                return result;
+            }
+        }
+
+        int _fileDescriptor;
+        GCHandle _handle;
+        Queue<TransferOperation> _receiveQueue;
+        Queue<TransferOperation> _sendQueue;
+        Queue<AcceptOrConnectOperation> _acceptOrConnectQueue;
+        object _closeLock = new object();
+        object _queueLock = new object();
+        SocketAsyncEngine _engine;
+        uint _registeredEvents;
+
+        public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
+        {
+            _fileDescriptor = fileDescriptor;
+            _engine = engine;
+
+            var events = Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT | Interop.libc.EPOLLRDHUP | Interop.libc.EPOLLET;
+
+            Interop.Error errorCode;
+            if (!_engine.TryRegister(this, _fileDescriptor, events, 0, ref _handle, out errorCode))
+            {
+                throw new Exception(string.Format("SocketAsyncContext: {0}", errorCode));
+            }
+
+            _registeredEvents = events;
+        }
+
+        public void Close()
+        {
+            lock (_closeLock)
+            lock (_queueLock)
```

Comment:
It'd be worth adding a comment where these locks are defined about the lock hierarchy that needs to be followed to avoid deadlocks.  And then in places where _closeLock is acquired, asserting that _queueLock isn't held (```Debug.Assert(!Monitor.IsEntered(_queueLock));```).